### PR TITLE
io_uring: async segment write path

### PR DESCRIPTION
**Pull request by sitano**
At least save 1 syscall on the segment write path. Don't wait for fsync.

Implementation:
- separate write uring queue for writes
- writes usually require 2 SQE per write (write+fsync)
- fsyncs are issued with DRAIN and not waited
- URing uses preallocated iovecs structs and keep Arcs for used bufs
- when submit queue exhausts write() will wait for minimum empty slots
- uring is used via my custom binding to liburing similar to https://github.com/sitano/liburing

URing.pwrite_all(buf, data, offset, fsync) implementation is as follows:
- take lock
- drain completion queue to retain free slots
- reserve required number of SQEs
- if not enought SQE stlots, wait (1 syscall), then drain
- build write
- build fsync if needed
- submit

ISSUE: https://github.com/spacejam/sled/issues/571

TODO:
- configuration
- sqpoll support
- CI for io_uring on linux

FUTURE:
- fixed buffers + buffer pool for write path
- SQPOLL?

### CODE REVIEW

**Comment by sitano** [comment]
Looking at the kernel source, it's clear that SQPOLL kthread just spin-loops over the timeout. So setting it up for segments write path does not make much sense as it does not involve a lot of syscalls. i.e. for default flush interval 200ms, its only 5 times a sec by 2 syscalls write and fsync. No big gain there.

A very heavy read app may benefit tho from it. Maybe recovery too. Especially as the recovery process heavily uses many small dense reads from page cache. Sometimes even from single page.
But it also can be easily mitigated by simple buffering if there is no any. I don't know how Rust pread implemented. So no need in io_uring there.

**Comment by spacejam** [comment]
This is really awesome work! Thank you for spending so much time getting this working :) I'm going to merge it in, and we can iterate on it more without worrying about conflicts

File: src/pagecache/io_uring.rs [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        for i in 0..size {
+            uring.free_slots.push(i);
+        }
+
+        Ok(uring)
+    }
+
+    unsafe fn drain_cqe(&mut self) -> Result<()> {
+        loop {
+            let mut cqe: *mut io_uring_cqe = std::mem::zeroed();
```
[the remark this answers is not in the document]
Comment (sitano) [reply src/pagecache/io_uring.rs]: thank you @quininer for the review! will update according to your comments.

File: src/pagecache/io_uring.rs [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+                let i = self.free_slots.pop().unwrap();
+                self.buf[i] = Some(data);
+                self.iovecs[i].iov_base =
+                    buf.as_mut_ptr() as *mut std::ffi::c_void;
+                self.iovecs[i].iov_len = buf.len();
+
+                io_uring_prep_writev(
+                    sqe,
+                    self.file.as_raw_fd(),
```
[the remark this answers is not in the document]
Comment (sitano) [reply src/pagecache/io_uring.rs]: good catch! missed it, even tho I register fd in the beginning. The next step will be using registered buffers.

File: src/pagecache/io_uring.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sitano) [reply src/pagecache/io_uring.rs]: it's done in master

File: src/pagecache/io_uring.rs [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+                        io::ErrorKind::Other,
+                        "wait for free sqes",
+                    )));
+                }
+            }
+
+            // 3. build write() SQE
+            {
+                let sqe = io_uring_get_sqe(&mut self.ring);
+                if sqe == std::ptr::null_mut() {
```
[the remark this answers is not in the document]
Comment (sitano) [reply src/pagecache/io_uring.rs]: it's done in master

File: src/pagecache/io_uring.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sitano) [reply src/pagecache/io_uring.rs]: fixed in #886

File: src/pagecache/io_uring.rs [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+                    (self.iovecs.len() - self.free_slots.len()) as u32,
+                );
+                if ret < 0 {
+                    panic!(Error::Io(io::Error::from_raw_os_error(ret)));
+                }
+            }
+
+            self.drain_cqe().unwrap();
+
+            io_uring_unregister_files(&mut self.ring);
```
[the remark this answers is not in the document]
Comment (sitano) [reply src/pagecache/io_uring.rs]: fixed in #886

**Pull request merged**